main.c

```diff
@@ -29,57 +33,95 @@ void printString (char* str) {
 }
 
 void saveCursorPos () {
-    printf("%c7", 27);
-    fflush(stdout);
+    printSpecialChar("7");
 }
 
 void restoreCursorPos () {
-    printf("%c8", 27);
-    fflush(stdout);
+    printSpecialChar("8");
+}
+
+void redrawPrevLines (char matrix[1000][1000], int currentLine)
+{
+  int i, j;
+  int width = w.ws_col;
+  saveCursorPos();
+  gotoxy(0, 0);
+  clearLine();
+  for (i = 1; i <= currentLine; i++){
+    gotoxy(0, i);
+    clearLine();
+    for(j = 1; j <= width; j++) {
+      gotoxy(j, i);
+      printf("%c", matrix[i][j]);
+    }
+  }
+  restoreCursorPos();
 }
 
 void *drawProgress(void *null)
 {
     int width = w.ws_col;
-    int height = w.ws_row;
+    // int height = w.ws_row;
     int n = 0;
     while (stop == 0) {
         usleep(20000);
         saveCursorPos();
-        gotoxy(n, height);
+        // gotoxy(n, height);
+        printSpecialChar("[1A");
+        printf("\r");
         n++;
         if (n > width - 1) {
             n = 0;
             clearLine();
         }
-        printString("#");
+        for(int i = 0; i < n; i++) printString("#");
         restoreCursorPos();
     }
     return NULL;
 
 }
 
+void *proceedInput (void *null)
+{
+  int c;
+  int width = w.ws_col;
+  int height = w.ws_row;
+  char charMatrix[1000][1000];
+  int currentLine = 1;
+  int n = 1;
```

Comment:
What is 'n' it's unclear from the context, except 'i' and 'j' please give more meaningful names.